### DISCUSSION TITLE
Integration with external security infrastructure

### DISCUSSION THREAD

**Discussion by MichaelMorrisEst** [strimzi/strimzi-kafka-operator#10081] [Ideas]
I have a scenario where I need to integrate Strimzi with an existing security infrastruture. All components must use certificates generated by the existing security infrastructure.
As Strimzi currently generates the certificates itself there is no option to integrate with other security infrastructure.

I would like to propose updating Strimzi to enable such integration, e.g. by Strimzi not generating the secrets contained the certs etc, but instead depending on the certs being created/updated by an external component. The default behaviour would continue to be as today, but the newly proposed behaviour shall be possible to enable through configuration.

Is this something the community is open to considering a proposal on?

**Comment by scholzj** [comment]
I guess it depends a lot on what your _existing security infrasturcture_ is and how would you imagine it to work. @katheris is currently looking into improving the support for integration with tools such as Cert Manager.

**Reply by MichaelMorrisEst** [reply]
The idea would be that the implementation in Strimzi would be agnostic of the security infrastructure, it would continue to use secrets as today, the only difference being, those secrets are generated externally to Strimzi rather than by Strimzi itself.

**Reply by scholzj** [reply]
I'm not sure how exactly you imagine it. I don't think Strimzi can be agnostic to it because working TLS infrastructure needs many different parts to fit in to work properly. E.g. the certificates need to have certain configurations etc.

**Reply by MichaelMorrisEst** [reply]
ok, yes, I guess what I meant to say was that it would be agnostic of what generated the certs. The certs would need to be created by the external security infrastructure in similar manner to what is already created today by Strimzi itself. 
My aim would be that there would be no impacts in Strimzi at the points of using the certs. The main impacts would really be to not generate the certs and associated secrets

**Reply by scholzj** [reply]
But that does not really work without proper automation. If you want to generate the server certificates yourself, you can do it today. Create the Secrets with the right certificates with the right settings before deploying the cluster and it will be used. But in reality, that is not feasible to do manually. So you need to have some APIs and automate the issuing of the certifictes.

**Reply by MichaelMorrisEst** [reply]
If I create the certs and secrets myself (either manually or in a automated way) wont the reconciliation of those secrets in the Strimzi code replace them with certs based on the internal CA cert?

If my understanding of Cert Manager is correct, then it also creates the certs as secrets, so I'm guessing the integration here will be to allow the certs be read from the secrets created by Cert Manager rather than secrets created internally by Strimzi. If my understanding is correct here this sounds very similar to what I am trying to achieve
Is there any issue or other doc describing what is proposed for Cert Manager integration?

**Reply by scholzj** [reply]
> If I create the certs and secrets myself (either manually or in a automated way) wont the reconciliation of those secrets in the Strimzi code replace them with certs based on the internal CA cert?

Well ... you obviously have to use custom CA as well for this which you use to sign the server certs. But the operator pattern is that the operator checks if there are any changes needed, and if not, then it doesn't change things. But as I said, the tricky part is to make sure that the certificates are correct and do not require any change.

> If my understanding of Cert Manager is correct, then it also creates the certs as secrets, so I'm guessing the integration here will be to allow the certs be read from the secrets created by Cert Manager rather than secrets created internally by Strimzi. If my understanding is correct here this sounds very similar to what I am trying to achieve

Sure, Cert Manager will provide the certificates to Strimzi in Secrets. But it provides also an API for requesting the certificates. So the expectation is that Strimzi requests the certificates through the API and Cert Manage provides them. Not that someone manually creates the certificates with Cert Manager.

**Reply by MichaelMorrisEst** [reply]
Yes, if I could use a custom CA then that would be fine, however that would need access to the CA private key, which is not available in the scenario.

For the cert manager case I guess responsibility for creating/renewing certs will still lie with Strimzi, it just uses cert manager to perform these tasks for it (by invoking the cert manager API). 

What I would like to propose is that its possible to make that responsibility be external to Strimzi, i.e. the external security infrastructure is responsible for creating/renewing certs and creating/reconciling the secrets (without any prompting from Strimzi). The Strimzi components can just use the certs from the secrets without having to handle the creation/renewal of the certs or reconciling the secrets (as its handled externally to Strimzi). The proposal would be that this behaviour is available as a config option, with no change in behaviour by default

**Reply by scholzj** [reply]
But that will not work any better than it works today. Strimzi has to _prompt for the certificates_ because otherwise you will not know how the certificate should look like. Kafka / Strimzi is multilayered application with many internal connections etc. It is not a simple webapp where you provide a file with whatever server certificate you have and that is it. So setting up the certificates to work is non-trivial.

**Reply by MichaelMorrisEst** [reply]
Yes, I have spent the last few weeks looking into where and how the certs are generated and used in the various components and have experimented with creating my own certs manually. I haven't come across anything that I see as a blocker. You need to know the domain names to put in the certs but thats all knowable upfront. Is there some specific setting in the certs that you are concerned about?
Thanks for all your feedback

**Reply by scholzj** [reply]
> Yes, I have spent the last few weeks looking into where and how the certs are generated and used in the various components and have experimented with creating my own certs manually. I haven't come across anything that I see as a blocker. You need to know the domain names to put in the certs but thats all knowable upfront. Is there some specific setting in the certs that you are concerned about?

That is not true. It will heavily depend on your configuration. You need to use:
* Compatible algorithms
* Have the right subjects 
* Have the right SANs

But for example, the SANs are dynamically determined based on your listeners and other configurations. So no, in many cases they are not known upfront. Issuing the certs manually also means problems when you scale things and so on. So it might be that it works _easily_ for you. But that does not mean it works for anyone else.

That said, if you tried it and it works nicely for you as you suggest, what exactly it is you are missing? I would assume that you can prevent the operator from overwriting your own certs by supplying some dummy (invalid) private key as part of your own CA and managing the certificate secrets your self.

**Reply by katheris** [reply]
Hey, I’ve been away at a conference, so just catching up. As Jakub said I am looking at adding support for certificate manager, however a large part of that work is around changing the way we deal with certificates internally, since as it stands today the assumption that Strimzi issues its own certificates is baked into the code everywhere. So it may be that once that refactoring work is done it would be easier to see how we would support an arbitrary certificate issuer. That being said, I am not aware of many users requesting this feature and I suspect it would be a lot of work to enable it in a way that was easy for users to use and worked without lots of caveats. Since as Jakub already noted there various certificate requirements that are determined by exactly how the Kafka CR is configured.

For the cert manager integration we are expecting that Strimzi will interact directly with cert manager to request certificates. So if we were to support this feature I think we would have to do it a similar way, so the user provides a class that we plugin to the Strimzi cluster operator that understands how to request certificates from the security infrastructure. That way Strimzi is still able to specify the SANs and subjects etc it requires for each certificate

**Reply by MichaelMorrisEst** [reply]
Yes the certs and the strimzi configuration need to match but I don't think that is difficult to achieve (but perhaps I'm missing something)

Going through the certs for some of the components as an example

The cert for the cluster operator has:
CN = cluster-operator
No SANs

The cert for the topic operator has:
CN = \<cluster name>-entity-topic-operator
No SANs

The certs for the kafka brokers have:
CN = \<cluster name>-kafka
SANs: 
\<cluster name>-kafka-brokers, 
\<cluster name>-kafka-brokers.\<namespace>, 
\<cluster name>-kafka-brokers.\<namespace>.svc, 
\<cluster name>-kafka-brokers.\<namespace>.svc.cluster.local

\<cluster name>-kafka-bootstrap, 
\<cluster name>-kafka-bootstrap.\<namespace>, 
\<cluster name>-kafka-bootstrap.\<namespace>.svc, 
\<cluster name>-kafka-bootstrap.\<namespace>.svc.cluster.local, 

\<node name>.\<cluster name>-kafka-brokers.\<namespace>.svc, 		
\<node name>.\<cluster name>-kafka-brokers.\<namespace>.svc.cluster.local, 
 with \<node name> = \<cluster name>-kafka-\<node id>

and also the bootstrap alternativeNames and brokers host from the listeners configuration


The only complication I see is the node id but this can be addressed using a wildcard

If I'm missing something then your insights would be appreciated


From a coding point of view the impacts are actually quiet straightforward, mostly just adding if statements to not execute the secret/cert handling code when this behaviour is wanted

**Reply by scholzj** [reply]
Yeah, you are missing all the dynamic names for the external listeners - for example based on Load Balancer IPs or DNS names. I think you also still did not explain what does not work on this today if it is as eas and simple as you suggest.

**Reply by MichaelMorrisEst** [reply]
Thanks for your suggestion to use a custom CA with a dummy private key, it solves part of the problem in that the CA is never renewed and the (externally created) secrets are used as provided rather than being overwritten.

Whats missing is as follows:

1. The secrets containing the certs for the operators, kafka nodes etc are continuously reconciled and, when the time left to the expiration date in the certs falls within the renewal period, Strimzi will attempt to renew the certs. This is a problem for this use case as the renewal should be left to the external actors. In mitigation, we could set the renewal period to the minimum of 1 day but it would be better to be able disable the attempted renewal. This could be achieved by allowing the renewal period be set to 0 or by some more explicit new config option similar to the existing 'generateCeritificateAuthority' (which is, in essence what I am proposing).
2. The code expects the kafka and zookeeper secrets will contain the certs for each pod, with secret key based on the node name (i.e. including the node id). This means in creating the secret externally you would need to add entries for the maximum possible number of pods. A better solution for this use case would be to also allow a key that can be used for all pods (containing a wildcard SAN, i.e. <cluster name>-kafka-*...)

**Reply by scholzj** [reply]
Re 1) I can imagine that some flag to disable the certificate management might be considered once the work around CertManager integration is done. I would expect that to have some pluggable mechanism so you might be for example able to provide some dummy _do nothing_ implementation. But I do not think it makes sense to introduce it right now when we expect this area to be reworked significantly.

But that basically changes only the error you get (you get a nicer error `You failed to renew the certificates` instead of some error that it failed to generate the new certs) and shifts it by one day max (because it will fail when the certs expire, exactly one day later if you set the renewal period to 1 day).

This is something what at least I personally absolutely do not want anyone to do (if nothing else then because nobody really wants to spend days trying to help some users lost in how to generate the proper certifciates and why is it not working with their wrong certificates). So I think for that, this is not a bad workaround.

Re 2) Funny. When I said how hard it is to want to do all of this manually, you seemed to insist that it is easy 😉. I do not think we plan to use a single certificate. Wildcard certificates are pretty limited and cannot be used for all the SANs. It might also run into problems with the number of SANs in a single certificate. If the wildcard certificate works for you, I think you can easily copy it multiple times into the Secret. Also, please keep in mind that in the future the certificates are expected to be split over multiple secrets and not all in one secret because of various limitations such as Secret size.

**Reply by MichaelMorrisEst** [reply]
Hi relation to the cert manager integration, is there a timeline in mind for that? I would be happy to help out with that work

**Reply by katheris** [reply]
It's in progress at the moment, although we don't have a specific timeline set out. So far I've done some refactoring of the internals to better align things and then I've recently raised a proposal to start creating better separation of concerns. The Strimzi code generally has assumptions built in about how the certificates are managed, so a large part of the work is starting to pick that apart.

That's great if you'd like to help out, a good starting point would be reviewing the [open proposal PR](https://github.com/strimzi/proposals/pull/116). I'm aiming to break down the work into manageable chunks so I'm sure we could find parts to parallelise, but they'll definitely be proposals and PRs that it would be good to get more eyes on either way. If you want some more context there is also this [closed proposal](https://github.com/strimzi/proposals/pull/46). It's a bit out of date, but it's still roughly the direction I'm aiming at.
